pkg/agent/openflow/network_policy.go

```diff
@@ -739,31 +758,61 @@ func (c *client) InstallPolicyRuleFlows(ruleID uint32, rule *types.PolicyRule, n
 		} else {
 			actionFlows = append(actionFlows, c.conjunctionActionFlow(ruleID, ruleTable.GetID(), dropTable.GetNext(), rule.Priority))
 		}
-		if err := c.ofEntryOperations.AddAll(actionFlows); err != nil {
-			return nil
-		}
-		// Add the action flows after the Openflow entries are installed on the OVS bridge successfully.
 		conj.actionFlows = actionFlows
 	}
-	c.conjMatchFlowLock.Lock()
-	defer c.conjMatchFlowLock.Unlock()
+	return conj
+}
 
+// calculateMatchFlowChangesForRule calculates the contextChanges for the policyRule, and updates the context status in case of batch install.
+func (c *client) calculateMatchFlowChangesForRule(conj *policyRuleConjunction, rule *types.PolicyRule, isBatchInstall bool) []*conjMatchFlowContextChange {
 	// Calculate the conjMatchFlowContext changes. The changed Openflow entries are included in the conjMatchFlowContext change.
 	ctxChanges := conj.calculateChangesForRuleCreation(c, rule)
+	// Update conjunctiveMatchContext if during batch flow install, otherwise the subsequent contextChange
+	// calculations will not be based on the previous flowChanges that have not been sent to OVS bridge.
+	// TODO: roll back if batch flow install fails?
+	if isBatchInstall {
+		for _, ctxChange := range ctxChanges {
+			ctxChange.updateContextStatus()
+		}
+	}
+	return ctxChanges
+}
 
-	// Send the changed Openflow entries to the OVS bridge, and then update the conjMatchFlowContext as the expected status.
-	if err := c.applyConjunctiveMatchFlows(ctxChanges); err != nil {
+// BatchInstallPolicyRuleFlows installs flows for NetworkPolicy rules in case of agent restart. It calculates and
+// accumulates all Openflow entry updates required and installs all of them on OVS bridge in one bundle.
+func (c *client) BatchInstallPolicyRuleFlows(ofPolicyRules []*types.PolicyRule) error {
+	c.replayMutex.RLock()
+	defer c.replayMutex.RUnlock()
+
+	var allCtxChanges []*conjMatchFlowContextChange
+	var allActionFlowChanges []*binding.Flow
+	var updatedConjunctions []*policyRuleConjunction
+
+	for _, rule := range ofPolicyRules {
+		conj := c.calculateActionFlowChangesForRule(rule)
+		ctxChanges := c.calculateMatchFlowChangesForRule(conj, rule, true)
+		for _, actionFlow := range conj.actionFlows {
+			allActionFlowChanges = append(allActionFlowChanges, &actionFlow)
+		}
```

Comment:
I guess the result in allActionFlowChanges would be unexpected as you are storing the address of the loop variable into another slice, which could result in all items pointing to the last item.
Check this: https://play.golang.org/p/nVBcBm0B09z
I think you don't need to turn the slice of interfaces into a slice of interface pointers. You can just copy the items.
If the issue is real, and it doesn't break any CI test, I guess we need a test case to cover it.

Reply:
Thanks for identifying the issue. Fixed and added a testcase.

Reply:
I think the previous code will cause some actions that are supposed to apply but are not. Is it because each rule has only one actionFlow in your test?
Just `allActionFlowChanges = append(allActionFlowChanges, actionFlow...)` like allCtxChanges below?

Reply:
As of now all rules have one actionFlow each by design, not only in tests.